VideoPress: Uploads Via Media Modal

dbtlr opened the pull request:
Currently, VideoPress uploads are only able to occur via the VideoPress tab. This is largely because of the complication with uploading media of large file sizes, which is difficult on most hosts, paired with the fact that the standard pluploader is not used, because it is a complicated flow to upload to 2 different locations.

This change is intended to not only add in a second path to the standard uploader, that when video files are detected, will send those files to WPCOM instead, but will also begin using a different type of upload token to authenticate that the user is allowed to upload.

Another side effect of this change is that it also introduces the idea of a Jetpack blog uploading to itself on the WPCOM API side, instead of the a WPCOM paired blog.

## Dependencies

- [x] WPCOM API endpoint updates deployed.

## Notes

Once this is deployed and verified, a secondary PR will be introduced to remove the old upload method. We also need to evaluate removing the site chooser as well, though there needs to be some discussion on moving videos from paired blogs to their Jetpack counterparts.

dbtlr commented:
As a note, this addresses issue #1952

jeherve commented:
I tried uploading a new VideoPress video via Add Media > VideoPress in the post editor, but it looks like it's still trying to upload to my local site:

![screen shot 2016-06-20 at 7 14 05 pm](https://cloud.githubusercontent.com/assets/426388/16203609/5177ce7e-371b-11e6-9434-6be475f6c641.png)

I then tried to upload via Media > VideoPress, but got the following error:

> Could not obtain a VideoPress upload token. Please try again later.

dbtlr commented:
@jeherve I'll take a look to see if I can replicate.

This should have no direct dependency on #4143 to handle the upload itself. However, without #4143, the media item won't be updated once the transcoding process finishes, so the total experience is incomplete without it.

jeherve commented:
@dbtlr Let me know if you want access to my site to test.

dbtlr commented:
@jeherve There is some sort of difference between how the production wpcom api in handling empty bodied jetpack requests and how our sandboxes work. I'm troubleshooting it now, though it is difficult to get proper debugging information with these types of problems unfortunately...

dbtlr commented:
Okay, found the issue with the API and fixed it.

As I suspected, there was an issue with a NULL vs "" in how jetpack signatures calculated the body_hash. For whatever reason the 2 environments were slightly different as to which was present. Not sure why, but in this case it is okay to assume the two are the same.

Also, you were correct that this depends on #4143. Once the upload succeeds (after the API was fixed) it will then need to call back to the Jetpack blog to create the media item. If this fails, the upload fails.

jeherve commented:
My Token trouble is now gone, but I still get the error when trying to upload via the new editor; it seems it's still trying to upload to my local site.

dbtlr commented:
@jeherve This should be fixed now. There was an issue with the default plupload max_file_size filter running in parallel to our uploader filter, so I merged the 2, which has fixed the issue.

georgestephanis commented on modules/videopress/videopress.php:
```diff
@@ -56,24 +56,47 @@ function on_init() {
-			return wp_send_json_error();
+		if ( ! $this->can( 'upload_videos' ) ) {
+			wp_send_json_error();
+			return;
```
We can probably skip the `return;` here -- it's fine to leave in for clarity's sake, but it will never run.  `wp_send_json()` -- which is called by `wp_send_json_error()` -- ends with either a `die;` or `wp_die();` depending on if it's an ajax request according to `DOING_AJAX`.

dbtlr replied:
it's more for clarity sake and in case the die were to ever change (as unlikely as that may be).

georgestephanis commented on modules/videopress/videopress.php:
```diff
@@ -56,24 +56,47 @@ function on_init() {
-		return wp_send_json_success( $response );
+		if ( empty( $response['upload_blog_id'] ) || empty( $response['upload_token'] ) ) {
+			wp_send_json_error( array( 'message' => __( 'Could not obtain a VideoPress upload token. Please try again later.', 'jetpack' ) ) );
+			return;
```
Same remark as on the `return;` line up above

georgestephanis commented on modules/videopress/videopress.php:
```diff
@@ -731,6 +759,33 @@ function videopress_shortcode_options( $options ) {
+			'%s://%s/rest/v%s/videos/%s/new',
+			'https',
+			JETPACK__WPCOM_JSON_API_HOST,
+			Jetpack_Client::WPCOM_JSON_API_VERSION,
```
We may want to hardcode this here -- just because the wpcom rest api version gets a bump doesn't necessarily mean that there's a new version of the endpoint, does it?

georgestephanis commented on modules/videopress/videopress.php:
```diff
@@ -731,6 +759,33 @@ function videopress_shortcode_options( $options ) {
+	protected function make_media_upload_path( $blog_id ) {
+		return sprintf(
+			'%s://%s/rest/v%s/videos/%s/new',
+			'https',
```
What's the point of sprintf'ing a hardcoded protocol?

dbtlr replied:
Mostly consistency with how this is done in other parts of Jetpack. In this case, we are making requests from the browser, so we need to require that the calls are made over https.

georgestephanis replied:
If it were a ternary or the like that toggled between http and https, I'd maybe understand it, but if it's always going to be https, I'd sooner see that just replacing the %s in the original string instead of making php do that work on every load.

dbtlr commented:
It seems like there may be an issue with the API allowing for large files on the new upload endpoint. I'm waiting for a response on how we can get this fixed. Will update when this is fixed.

dbtlr commented:
The issues with the NGINX configuration has been resolved. :)

jeherve commented:
@dbtlr Do the video uploads work for you now? I tried, and I seem to get the following response for all uploads:

`{"error":"jetpack_response_error","message":"The Jetpack site encountered an error and could not process the API request."}`

This happens after I applied #4142 and #4144.

jeherve commented:
I'm still experiencing issues after applying #4142 and #4144, but the error message is now different:

`{"errors":[{"file":"2016-05-28 15.07.23.mp4","error":"upload_error","message":"Sorry, this file type is not permitted for security reasons."}]}`

@georgestephanis Could you take a look?

dbtlr commented:
That is a new error from the server. It looks like it is failing a mime type check deep within wpcom. Can you send me that file so I can do some tests as to why it might be failing?

jeherve commented:
@dbtlr I was able to upload it via the linked WordPress.com site's Media Library. The video GUID is `yK8pnoZe`.

Here is [the original file](https://videos.files.wordpress.com/yK8pnoZe/2016-05-28-15-07-23.mp4), if you want to try on your end.

dbtlr commented:
I've done a pretty far-reaching overhaul of the approach to these uploads since my last note.

Essentially the problem I ran into is that no matter how hard I tried, if the media item is created by the API asking the Jetpack blog to create it, there is a strong chance that when that returns the wpcom backend will not have the attachment synced to it yet. Because of that, I've switched my approach and we now create a new attachment PRIOR to uploading that is then being passed to the wpcom backend. This approach was picked to allow us the ability to have the sync occur from Jetpack to wpcom while the upload is being performed, to gain better accuracy of the sync.

I'm still seeing a few sync related errors where the media item has not been synced up to wpcom yet, this especially happens when the file size is small and the upload is fast. I've added some retries into the wpcom back end that attempt to wait for the sync to occur, however, this is a stop gap while I get a little more familiar with the sync system itself (which has an overhaul underway as well).

This approach has left me with a problem, though, in that media items are created that can potentially never have data filled into them. For that reason, they all have a custom `videopress_status` attachment that I plan on looking through occasionally and cleaning them up.

While I want to add that in prior to merge I'd love it if @jeherve could give this another go and see if some of the random failures might be resolved. As it stands right now, this plus #4142 are consistently working for me.

I'll get the cleanup method added into this PR ASAP.

rcoll commented:
Is this wired up to allow uploads from the Library screen (upload.php)? When dragging/dropping a video file onto this screen, it seems to upload but does not appear in the library on the Jetpack site nor on the connected WPCOM blog. This would be highly confusing for some users who don't typically use media-new.php for uploading.

mrjarbenne commented:
We are building something similar using the Vimeo API. Solving this might simplify things: https://core.trac.wordpress.org/ticket/35310

dbtlr commented:
This PR is being closed in favor of: #5326
